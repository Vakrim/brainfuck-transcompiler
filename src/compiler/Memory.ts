import { Address } from './Address';

export class Memory {
  #tape: Map<Address, ScopeLike>;

  constructor() {
    this.#tape = new Map();
  }

  allocate(scope: ScopeLike, nextTo: Address, size: number): Address {
    const address = this.#findFreeAdrress(nextTo, size);

    for (let i = 0; i < size; i++) {
      this.#tape.set((address + i) as Address, scope);
    }

    return address;
  }

  free(scope: ScopeLike, address: Address) {
    const cell = this.#tape.get(address);
    if (!cell) {
      throw new Error(`Cell with address = ${address} is not allocated`);
    }
    if (cell !== scope) {
      throw new Error(
        `Cell with address = ${address} is not allocated by the scope that is requesting to free it`
      );
    }

    this.#tape.delete(address);
  }

  #findFreeAdrress(nextTo: Address, size: number) {
    for (let searchRange = 0; ; searchRange++) {
      for (let i = 0; i < size; i++) {
        if (
          nextTo - searchRange - i < 0 ||
          this.#tape.has((nextTo - searchRange - i) as Address)
        ) {
          break;
        }
        if (i >= size - 1) {
          return (nextTo - searchRange - size + 1) as Address;
        }
      }

      for (let i = 0; i < size; i++) {
        if (this.#tape.has((nextTo + searchRange + i) as Address)) {
          break;
        }
        if (i >= size - 1) {
          return (nextTo + searchRange) as Address;
        }
      }
    }
  }
}

type ScopeLike = object;
