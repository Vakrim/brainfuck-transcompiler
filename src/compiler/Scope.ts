import { Address } from './Address';
import { Memory } from './Memory';
import { ArraySize, TemporaryArray } from './TemporaryArray';
import { TemporaryVariable } from './TemporaryVariable';
import { Variable } from './Variable';

export class Scope {
  #parent: Scope | null;
  #variables: Map<string, Variable>;
  #temporaries: Set<TemporaryVariable | TemporaryArray>;
  #memory: Memory;

  constructor(parent: Scope | null, memory: Memory) {
    this.#parent = parent;
    this.#memory = memory;
    this.#variables = new Map();
    this.#temporaries = new Set();
  }

  declareVariable(name: string, nextTo: Address) {
    if (this.#variables.has(name)) {
      throw new Error(`Can't redeclare variable with name "${name}"`);
    }

    const address = this.#memory.allocate(this, nextTo, 1);

    const variable = new Variable(name, address, this);

    this.#variables.set(name, variable);
  }

  unsetVariable(name: string) {
    if (!this.#variables.has(name)) {
      throw new Error(`Variable "${name}" is not declared`);
    }

    const variable = this.#variables.get(name)!;

    this.#variables.delete(name);

    this.#memory.free(this, variable.address);
  }

  declareTemporaryVariable(nextTo: Address) {
    const address = this.#memory.allocate(this, nextTo, 1);

    const variable = new TemporaryVariable(address);

    this.#temporaries.add(variable);

    return variable;
  }

  declareTemporaryArray<T extends ArraySize>(nextTo: Address, size: T) {
    const address = this.#memory.allocate(this, nextTo, size);

    const array = new TemporaryArray(address, size);

    this.#temporaries.add(array);

    return array;
  }

  unsetTemporaryVariable(variable: TemporaryVariable) {
    if (!this.#temporaries.has(variable)) {
      throw new Error(`Temporary variable is not declared`);
    }

    this.#temporaries.delete(variable);

    this.#memory.free(this, variable.address);
  }

  unsetTemporaryArray(array: TemporaryArray) {
    if (!this.#temporaries.has(array)) {
      throw new Error(`Temporary array is not declared`);
    }

    this.#temporaries.delete(array);

    array.variables.forEach((variable) => {
      this.#memory.free(this, variable.address);
    });
  }

  getVariable(name: string): Variable {
    return this.getScopeOfVariable(name).#variables.get(name)!;
  }

  getScopeOfVariable(name: string): Scope {
    if (this.#variables.has(name)) {
      return this;
    }

    if (!this.#parent) {
      throw new Error(`Variable "${name}" doesn't exist`);
    }

    return this.#parent.getScopeOfVariable(name);
  }

  getVariablesOfThisScope(): string[] {
    return Array.from(this.#variables.keys());
  }

  getParentScope(): Scope {
    if (!this.#parent) {
      throw new Error('There are not parent Scope');
    }

    return this.#parent;
  }

  hasParentScope(): boolean {
    return !!this.#parent;
  }

  deepVerifyBeforeDiscard() {
    this.verifyBeforeDiscard();

    if (this.#parent) {
      this.#parent.deepVerifyBeforeDiscard();
    }
  }

  verifyBeforeDiscard() {
    if (this.#temporaries.size > 0) {
      throw new Error(
        `Can't discard scope when temporary variables are still allocated`
      );
    }
  }

  get memory() {
    return this.#memory;
  }
}
