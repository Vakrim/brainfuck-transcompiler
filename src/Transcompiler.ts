import { Address } from './Address';
import { CodePrinter, CodeOperation } from './CodePrinter';
import { Memory } from './Memory';
import { scopeSymbol } from './MemoryAllocationSnap';
import { Scope } from './Scope';
import { TemporaryVariable } from './TemporaryVariable';
import { Variable } from './Variable';

export class Transcompiler {
  #cursorPosition: Address = 0 as Address;
  #memory: Memory;
  #scope: Scope;
  #code: CodeOperation[];
  #currentCodeBlock: string;
  #codeBlockNamesStack: string[];

  constructor(memory = new Memory()) {
    this.#cursorPosition = 0 as Address;
    this.#memory = memory;
    this.#scope = new Scope(null, this.#memory);
    this.#code = [];
    this.#currentCodeBlock = '';
    this.#codeBlockNamesStack = [];
  }

  declareVariable(name: string) {
    this.#operation(`declare ${name}`, () => {
      this.#scope.declareVariable(name, this.#cursorPosition);
    });
  }

  assignValue(name: string, value: number) {
    this.#operation(`assign ${value} to ${name}`, () => {
      this.#reset(name);
      if (value > 0) {
        this.#inc(name, value);
      } else {
        this.#dec(name, -value);
      }
    });
  }

  writeInput(name: string) {
    this.#operation(`write input ${name}`, () => {
      this.#moveTo(name);
      this.#outputBrainfuck(',');
    });
  }

  printVariable(name: string) {
    this.#operation(`print ${name}`, () => {
      this.#printValue(name);
    });
  }

  printNumber(name: string) {
    this.#operation(`print ${name}`, () => {
      const ten = this.#declareTemporaryVariable(name);
      const tens = this.#declareTemporaryVariable(name);
      const ones = this.#declareTemporaryVariable(name);

      this.#inc(ten, 10);

      this.#divmod(tens, ones, name, ten);

      this.#reset(ten);
      this.#scope.unsetTemporaryVariable(ten);

      this.whenever(tens, () => {
        this.#inc(tens, 48);
        this.#printValue(tens);
        this.#reset(tens);
      });

      this.#inc(ones, 48);
      this.#printValue(ones);
      this.#reset(ones);

      this.#scope.unsetTemporaryVariable(tens);
      this.#scope.unsetTemporaryVariable(ones);
    });
  }

  add(to: string, from: string) {
    this.#operation(`add ${from} to ${to}`, () => {
      this.#add(to, from);
    });
  }

  increment(name: string, n: number = 1) {
    this.#operation(`increment ${name} with ${n}`, () => {
      this.#inc(name, n);
    });
  }

  decrement(name: string, n: number = 1) {
    this.#operation(`decrement ${name} with ${n}`, () => {
      this.#dec(name, n);
    });
  }

  whenever(
    conditionName: Addressable,
    ifPostitive: () => void,
    ifNegative?: () => void
  ) {
    const conditionCopy = this.#copy(conditionName);

    if (!ifNegative) {
      this.#operation(`if ${this.#commentVariable(conditionName)}`, () => {
        this.#moveTo(conditionCopy);
        this.#outputBrainfuck(`[`);
      });
      this.scope(ifPostitive);
      this.#operation(`endif ${this.#commentVariable(conditionName)}`, () => {
        this.#moveTo(conditionCopy);
        this.#outputBrainfuck(`[-]]`);
      });
    } else {
      const elseFlag = this.#declareTemporaryVariable(conditionCopy);

      this.#operation(`if ${this.#commentVariable(conditionName)}`, () => {
        this.#inc(elseFlag);
        this.#moveTo(conditionCopy);
        this.#outputBrainfuck(`[`);
      });
      this.scope(ifPostitive);
      this.#operation(`else ${this.#commentVariable(conditionName)}`, () => {
        this.#reset(conditionCopy);
        this.#dec(elseFlag);
        this.#moveTo(conditionCopy);
        this.#outputBrainfuck(`]`);
        this.#moveTo(elseFlag);
        this.#outputBrainfuck(`[`);
      });
      this.scope(ifNegative);
      this.#operation(`endif ${this.#commentVariable(conditionName)}`, () => {
        this.#dec(elseFlag);
        this.#outputBrainfuck(`]`);
        this.#scope.unsetTemporaryVariable(elseFlag);
      });
    }

    this.#scope.unsetTemporaryVariable(conditionCopy);
  }

  multiply(to: string, from: string) {
    this.#operation(`multiply ${from} to ${to}`, () => {
      const interator = this.#copy(from);
      const interator2swap = this.#declareTemporaryVariable(interator.address);
      const result = this.#declareTemporaryVariable(interator.address);

      this.#loopOf(interator, () => {
        this.#loopOf(to, () => {
          this.#inc(interator2swap);
          this.#inc(result);
        });
        this.#loopOf(interator2swap, () => {
          this.#inc(to);
        });
      });

      this.#scope.unsetTemporaryVariable(interator);
      this.#reset(interator2swap);
      this.#scope.unsetTemporaryVariable(interator2swap);

      this.#reset(to);
      this.#restoreValueFromTemporary(to, result);
    });
  }

  divmod(
    divResult: string,
    modResult: string,
    dividentOriginal: string,
    divisorOriginal: string
  ) {
    this.#operation(
      `divmod (${divResult} ${modResult}) = ${dividentOriginal} / ${divisorOriginal}`,
      () => {
        this.#divmod(divResult, modResult, dividentOriginal, divisorOriginal);
      }
    );
  }

  isGreaterThanOrEqual(result: string, originalA: string, originalB: string) {
    this.#operation(`${result} = ${originalA} >= ${originalB}`, () => {
      const a = this.#copy(originalA);
      const b = this.#copy(originalB);
      this.#reset(result);

      this.whenever(
        b,
        () => {},
        () => {
          this.#inc(result);
        }
      );

      this.#loopOf(a, () => {
        this.#dec(b);
        this.whenever(
          b,
          () => {},
          () => {
            this.#inc(result);
          }
        );
      });

      this.#reset(b);
      this.#scope.unsetTemporaryVariable(b);
      this.#scope.unsetTemporaryVariable(a);
    });
  }

  while(name: string, fn: () => void) {
    this.#operation(`while ${name}`, () => {
      this.#moveTo(name);
      this.#outputBrainfuck(`[`);
    });
    this.scope(fn);
    this.#operation(`endwhile ${name}`, () => {
      this.#moveTo(name);
      this.#outputBrainfuck(`]`);
    });
  }

  scope(fn: () => void) {
    this.#pushScope();
    fn();
    this.#popScope();
  }

  get code(): string {
    this.#scope.deepVerifyBeforeDiscard();

    this.#operation('', () => {});

    return new CodePrinter(this.#code).print();
  }

  #operation(name: string, fn: () => void) {
    if (this.#codeBlockNamesStack.length > 0) {
      this.#code.push({
        name: this.#codeBlockNamesStack[this.#codeBlockNamesStack.length - 1],
        code: this.#currentCodeBlock,
        level: this.#codeBlockNamesStack.length,
      });
      this.#currentCodeBlock = '';
    }

    this.#codeBlockNamesStack.push(name);
    fn();
    if (this.#currentCodeBlock) {
      this.#code.push({
        name,
        code: this.#currentCodeBlock,
        level: this.#codeBlockNamesStack.length,
      });
      this.#currentCodeBlock = '';
    }
    this.#codeBlockNamesStack.pop();
  }

  #restoreValueFromTemporary(
    target: Addressable,
    temporaryWithValue: TemporaryVariable
  ) {
    this.#moveValue(target, temporaryWithValue);
    this.#scope.unsetTemporaryVariable(temporaryWithValue);
  }

  #add(to: Addressable, from: Addressable) {
    this.#moveTo(from);
    const newFrom = this.#declareTemporaryVariable(this.#cursorPosition);

    this.#loopOf(from, () => {
      this.#inc(newFrom);
      this.#inc(to);
    });

    this.#restoreValueFromTemporary(from, newFrom);
  }

  #divmod(
    divResult: Addressable,
    modResult: Addressable,
    dividentOriginal: Addressable,
    divisorOriginal: Addressable
  ) {
    const array = this.#scope.declareTemporaryArray(
      (this.#normalizeAddress(divResult) + 2) as Address,
      7
    );

    this.#add(array.at(0), dividentOriginal);
    this.#add(array.at(2), divisorOriginal);

    this.#moveTo(array.at(0));
    this.#outputBrainfuck(`[->+>-[>+>>]>[+[-<+>]>+>>]<<<<<<]`);

    this.#reset(modResult);
    this.#moveValue(modResult, array.at(3));
    this.#reset(divResult);
    this.#moveValue(divResult, array.at(4));

    array.variables.forEach((v) => this.#reset(v));

    this.#scope.unsetTemporaryArray(array);
  }

  #pushScope() {
    this.#code.push({
      scope: 'open',
    });
    this.#scope = new Scope(this.#scope, this.#memory);
  }

  #popScope() {
    this.#scope.verifyBeforeDiscard();

    const variables = this.#scope.getVariablesOfThisScope();
    for (const variable of variables) {
      this.#reset(variable);
      this.#scope.unsetVariable(variable);
    }
    this.#scope = this.#scope.getParentScope();
    this.#code.push({
      scope: 'close',
    });
  }

  #declareTemporaryVariable(nextTo: Addressable = this.#cursorPosition) {
    const nextToAddress = this.#normalizeAddress(nextTo);
    const variable = this.#scope.declareTemporaryVariable(nextToAddress);
    return variable;
  }

  #copy(from: Addressable) {
    const newFrom = this.#declareTemporaryVariable();
    const newTo = this.#declareTemporaryVariable();

    this.#operation(
      `copy ${this.#commentVariable(from)} to ${this.#commentVariable(newTo)}`,
      () => {
        this.#loopOf(from, () => {
          this.#inc(newFrom);
          this.#inc(newTo);
        });

        this.#restoreValueFromTemporary(from, newFrom);
      }
    );

    return newTo;
  }

  #moveValue(to: Addressable, from: TemporaryVariable) {
    this.#loopOf(from, () => {
      this.#inc(to);
    });
  }

  #reset(name: Addressable) {
    this.#moveTo(name);
    this.#outputBrainfuck('[-]');
  }

  #normalizeVariable(variableLike: VariableLike) {
    if (typeof variableLike === 'string') {
      return this.#scope.getVariable(variableLike);
    }
    return variableLike;
  }

  #normalizeAddress(addressable: Addressable) {
    if (typeof addressable === 'number') {
      return addressable;
    }
    return this.#normalizeVariable(addressable).address;
  }

  #moveTo(addressable: Addressable) {
    const address = this.#normalizeAddress(addressable);

    const difference = address - this.#cursorPosition;
    if (difference < 0) {
      this.#outputBrainfuck('<'.repeat(-difference));
    } else if (difference > 0) {
      this.#outputBrainfuck('>'.repeat(difference));
    }
    this.#cursorPosition = address;
  }

  #loopOf(iterator: Addressable, fn: () => void) {
    this.#moveTo(iterator);
    this.#outputBrainfuck(`[`);
    this.#dec(iterator);
    fn();
    this.#moveTo(iterator);
    this.#outputBrainfuck(`]`);
  }

  #outputBrainfuck(bf: string) {
    this.#currentCodeBlock += `${bf}`;
  }

  #dec(addressable: Addressable, value: number = 1) {
    this.#moveTo(addressable);
    this.#outputBrainfuck(`-`.repeat(value));
  }

  #inc(addressable: Addressable, value: number = 1) {
    this.#moveTo(addressable);
    this.#outputBrainfuck(`+`.repeat(value));
  }

  #printValue(addressable: Addressable) {
    this.#moveTo(addressable);
    this.#outputBrainfuck('.');
  }

  #commentVariable(addressable: Addressable) {
    if (typeof addressable === 'string') {
      return addressable;
    }
    if (addressable instanceof Variable) {
      return addressable.name;
    }
    const address = this.#normalizeAddress(addressable);
    return `temporary[${address}]`;
  }

  get [scopeSymbol]() {
    return this.#scope;
  }
}

type VariableLike = string | Variable | TemporaryVariable;

type Addressable = VariableLike | Address;
