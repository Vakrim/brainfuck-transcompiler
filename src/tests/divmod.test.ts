import { Transcompiler } from '../compiler/Transcompiler';
import { executeCode } from './execute-code';

describe('divmod', () => {
  it('calculates div and mod results', () => {
    const compiler = new Transcompiler();

    compiler.declareVariable('a');
    compiler.declareVariable('b');
    compiler.declareVariable('mod');
    compiler.declareVariable('div');

    compiler.writeInput('a');
    compiler.writeInput('b');

    compiler.divmod('div', 'mod', 'a', 'b');

    compiler.printVariable('div');
    compiler.printVariable('mod');

    expect(compiler.code).toMatchSnapshot();

    expect(executeCode(compiler, [24, 10]).memoryAllocation).toEqual([
      { name: 'a', value: 24 },
      { name: 'b', value: 10 },
      { name: 'mod', value: 4 },
      { name: 'div', value: 2 },
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ]);

    expect(executeCode(compiler, [24, 10]).codes).toEqual([2, 4]);
  });
});
