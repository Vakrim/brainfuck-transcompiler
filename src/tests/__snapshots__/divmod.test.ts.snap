// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`divmod calculates div and mod results 1`] = `
",  write input a
>,  write input b
<[->>>>+>+<<<<<]>>>>[-<<<<+>>>>]<<<[->>>+>>>+<<<<<<]>>>[-<<<+>>>]>[->+>-[>+>>]>[+[-<+>]>+>>]<<<<<<]<<<[-]>>>>>>[-<<<<<<+>>>>>>]<<<<<[-]>>>>>>[-<<<<<<+>>>>>>]<<<<[-]>[-]>[-]>[-]>[-]>[-]>[-]  divmod (div mod) = a / b
<<<<<<<<.  read div
<.  read mod"
`;
